controller/controller_server/csi_controller_server.py

```diff
@@ -132,6 +141,70 @@ def CreateVolume(self, request, context):
+        vol_name = vol.volume_name
+        try:
+            src_snapshot = array_mediator.get_snapshot_by_id(src_snapshot_id)
+            if not src_snapshot:
+                raise controller_errors.SnapshotNotFoundError(src_snapshot_id)
+            src_snapshot_name = src_snapshot.snapshot_name
+            src_snapshot_capacity = src_snapshot.capacity_bytes
+            logger.debug("Copy snapshot {0} data to volume {1}.".format(src_snapshot_id, vol_name))
+            array_mediator.copy_to_existing_volume_from_snapshot(vol_name, src_snapshot_name, src_snapshot_capacity,
+                                                                 min_vol_size)
+            logger.debug("Copy volume from snapshot finished")
+        except controller_errors.VolumeNotFoundError as ex:
+            logger.error("Volume not found while copying snapshot data to volume")
+            logger.exception(ex)
+            raise ex
+        except Exception as ex:
+            logger.error("Exception raised while copying snapshot data to volume")
+            self._rollback_create_volume_from_snapshot(array_mediator, vol.id)
+            raise ex
+        return vol
+
+    @retry(Exception, tries=5, delay=1)
+    def _rollback_create_volume_from_snapshot(self, array_mediator, vol_id):
+        logger.debug("Rollback copy volume from snapshot. Deleting volume {0}".format(vol_id))
+        array_mediator.delete_volume(vol_id)
+
+    def _handle_existing_vol_src_snap(self, volume, src_snapshot_id, context):
+        """
+        Args:
+            volume             : volume fetched or created in CreateVolume
+            src_snapshot_id    : id of snapshot we should copy to vol or None if volume should not be copied
+            context            : CreateVolume response context
+        Returns:
+            If volume exists and is a copy of specified snapshot - set context status to OK
+            and return CreateVolumeResponse.
+            If volume is a copy of another source - set context status to INTERNAL and return CreateVolumeResponse.
+            In any other case return None.
+        """
+        vol_name = volume.volume_name
+        vol_copy_src_object_id = volume.copy_src_object_id
+        if not src_snapshot_id or not vol_copy_src_object_id:
+            return None
+        if vol_copy_src_object_id == src_snapshot_id:
+            logger.debug(
+                "Volume {0} exists and it is a copy of snapshot {1}.".format(vol_name, src_snapshot_id))
+            context.set_code(grpc.StatusCode.OK)
+            return csi_pb2.CreateVolumeResponse()
+        else:
+            logger.debug(
+                "Volume {0} exists but it is not a copy of snapshot {1}.".format(vol_name, src_snapshot_id))
+            context.set_details("Volume already exists but it was created from a different source.")
+            context.set_code(grpc.StatusCode.INTERNAL)
```

Comment:
according to the [spec](https://github.com/container-storage-interface/spec/blob/v1.2.0/spec.md#createvolume):
"If a volume corresponding to the specified volume `name` already exists, is accessible from `accessibility_requirements`, and is compatible with the specified `capacity_range`, `volume_capabilities` and `parameters` in the `CreateVolumeRequest`, the Plugin MUST reply `0 OK` with the corresponding `CreateVolumeResponse`."

and since `volume_content_source` is not one of name/capacity/capabilities/parameters, we should return `OK` (and not `INTERNAL`)

Reply:
I don't think OK is correct. OK means all created or already exists as required.
Maybe they just didn't consider snapshots when adding this description. 
If you think it needs to be changed open a separate ticket.

Reply:
I think it would be better if we comply with the spec, instead of assuming the spec is wrong.
I opened ticket CSI-1540